cudax/samples/vector_add/vector_add.cu

```diff
@@ -0,0 +1,202 @@
+{
+  return std::forward<Arg>(arg);
+}
+
+template <typename Arg>
+using __prelaunch_t = decltype(__prelaunch(std::declval<stream_ref>(), std::declval<Arg>()));
+
+struct __fn
+{
+  template <typename Arg>
+  __prelaunch_t<Arg> operator()(stream_ref stream, Arg&& arg) const noexcept
+  {
+    // Intentionally unqualified call to __prelaunch to allow ADL
+    return __prelaunch(stream, std::forward<Arg>(arg));
+  }
+};
+} // namespace __prelaunch_
+
+inline constexpr __prelaunch_::__fn __prelaunch{};
+
+template <typename... ExpArgs, typename... ActArgs, typename... Levels>
+void launch_ex(
+  ::cuda::stream_ref stream, const hierarchy_dimensions<Levels...>& dims, void (*kernel)(ExpArgs...), ActArgs&&... args)
+{
+  static_assert(sizeof...(ExpArgs) == sizeof...(ActArgs), "Number of expected and actual arguments must match");
+
+  cudaError_t status = [&](auto&&... args) {
+    return detail::launch_impl(
+      stream, kernel_config(dims), kernel, static_cast<ExpArgs>(static_cast<decltype(args)>(args))...);
+  }(__prelaunch(stream, std::forward<ActArgs>(args))...);
+
+  if (status != cudaSuccess)
+  {
+    ::cuda::__throw_cuda_error(status, "Failed to launch a kernel");
+  }
+}
+} // namespace cuda::experimental
+
+/**
+ * CUDA Kernel Device code
+ *
+ * Computes the vector addition of A and B into C. The 3 vectors have the same
+ * number of elements numElements.
+ */
+__global__ void vectorAdd(cudax::span<const float> A, cudax::span<const float> B, cudax::span<float> C)
+{
+  int i = blockDim.x * blockIdx.x + threadIdx.x;
+
+  if (i < A.size())
+  {
+    C[i] = A[i] + B[i] + 0.0f;
+  }
+}
+
+/**
+ * Host main routine
+ */
+int main(void)
+{
+  // A CUDA stream on which to execute the vector addition kernel
+  cudax::stream stream;
+
+  // Print the vector length to be used, and compute its size
+  int numElements = 50000;
+  printf("[Vector addition of %d elements]\n", numElements);
+
+  // Allocate the host vectors
+  cudax::host_vector<float> h_A(numElements); // input
+  cudax::host_vector<float> h_B(numElements); // input
+  cudax::host_vector<float> h_C(numElements); // output
+
+  // Initialize the host input vectors
+  for (int i = 0; i < numElements; ++i)
+  {
+    h_A[i] = rand() / (float) RAND_MAX;
+    h_B[i] = rand() / (float) RAND_MAX;
+  }
+
+  // Allocate the device vectors
+  cudax::device_vector<float> d_A(numElements); // input
+  cudax::device_vector<float> d_B(numElements); // input
+  cudax::device_vector<float> d_C(numElements); // output
+
+  // Copy the host input vectors A and B in host memory to the device input
+  // vectors in device memory
+  printf("Copy input data from the host memory to the CUDA device\n");
+  d_A = h_A;
+  d_B = h_B;
+
+  // Define the kernel launch parameters
+  constexpr int threadsPerBlock = 256;
+  int blocksPerGrid             = (numElements + threadsPerBlock - 1) / threadsPerBlock;
+  auto dims = cudax::make_hierarchy(cudax::grid_dims(blocksPerGrid), cudax::block_dims<threadsPerBlock>());
```

Comment:
With #2001 merged we could use at_least(numElements) for the grid dimensions.
We could also try to come up with some shorthands for the entire hierarchy dimensions like that, they are super common

Reply:
how about this for a shorthand:
```c++
auto dims = cudax::distribute<256>(numElements);
```

Reply:
I like that